[DependencyInjection] Make method (setter) autowiring configurable

dunglas opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | maybe? |
| Tests pass? | yes |
| Fixed tickets | #19631 |
| License | MIT |
| Doc PR | symfony/symfony-docs#7041 |

Follow up of #19631. Implements https://github.com/symfony/symfony/pull/19631#issuecomment-240646169:

Edit: the last supported format:

``` yaml
services:
    foo:
        class: Foo
        autowire: ['__construct', 'set*'] # Autowire constructor and all setters
        autowire: true # Converted by loaders in `autowire: ['__construct']` for BC
        autowire: ['foo', 'bar'] # Autowire only `foo` and `bar` methods
```

Outdated:

``` yaml
autowire: true # constructor autowiring
autowire: [__construct, setFoo, setBar] # autowire whitelisted methods only
autowire: '*' # autowire constructor + every setters (following existing rules for setters autowiring)
```
- [x] Allow to specify the list of methods in the XML loader
- [x] Add tests for the YAML loader

dunglas committed:
[DependencyInjection] Introduce method injection for autowiring

dunglas commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -662,13 +662,28 @@ public function setAutowiringTypes(array $types)
      */
     public function isAutowired()
```
Should I deprecate this method?

fabpot replied:
Not sure it's worth it.

fabpot commented:
👍

dunglas commented:
The syntax for the XML loader is:

```xml
<?xml version="1.0" encoding="utf-8"?>
<container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
    <services>
        <service id="autowire_star" class="Foo" autowire="*" />
        <service id="autowire_array" class="Foo">
            <autowire>setFoo</autowire>
            <autowire>bar</autowire>
        </service>
    </services>
</container>
```

Does it looks good to everybody?

Ener-Getick commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -97,16 +98,32 @@ private function completeDefinition($id, Definition $definition)
+            $methods = array();
+            foreach ($autowired as $methodName) {
+                if ($reflectionClass->hasMethod($methodName)) {
+                    $methods[] = $reflectionClass->getMethod($methodName);
```
This won't work when using `['__construct']` right?

dunglas replied:
It works fine: https://3v4l.org/Lr78t

Ener-Getick replied:
Sure but it will add its arguments as a call instead of arguments. And it may create weird bugs and not work if there are mandatory arguments.

chalasr commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -41,8 +41,8 @@ public function process(ContainerBuilder $container)
             foreach ($container->getDefinitions() as $id => $definition) {
-                if ($definition->isAutowired()) {
-                    $this->completeDefinition($id, $definition);
+                if (false !== $autowired = $definition->getAutowired()) {
```
What about keep using `$definition->isAutowired()` here (as it stay valid), remove the newly added `$autowired` argument from `completeDefinition()` and use `$autowired = $defintion->getAutowired();` directly into?

dunglas replied:
It adds some extra method calls for no reason. I prefer to keep it as is.

chalasr replied:
The current adds an extra argument to `completeDefinition` for no reason, it can be done in the method directly, and the check is doing exactly what the method call would do thus ... Up to you.

Ener-Getick commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -72,8 +72,10 @@ public static function createResourceForClass(\ReflectionClass $reflectionClass)
 
-        foreach (self::getSetters($reflectionClass) as $reflectionMethod) {
-            $metadata[$reflectionMethod->name] = self::getResourceMetadataForMethod($reflectionMethod);
+        foreach ($reflectionClass->getMethods(\ReflectionMethod::IS_PUBLIC) as $reflectionMethod) {
```
Maybe we can allow masks? For example `set*`.

dunglas replied:
We can indeed allow masks in the configuration, but it's not related to this code block (it is not aware of the configuration so it needs to watch all methods anyway).

dunglas commented:
ping @symfony/deciders. We need to choose between merging this one or reverting setter autowiring before the 3.2 release.

Tobion commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -97,18 +100,65 @@ private function completeDefinition($id, Definition $definition)
+     */
+    public function getMethods(\ReflectionClass $reflectionClass, array $autowired)
+    {
+        if (is_array($autowired)) {
```
It's always an array as it is typehinted

Tobion commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -97,18 +100,65 @@ private function completeDefinition($id, Definition $definition)
+     *
+     * @return \ReflectionMethod[]
+     */
+    public function getMethods(\ReflectionClass $reflectionClass, array $autowired)
```
I think this is meant to be private

Tobion commented:
The loader should probably forbid to use both the autowire attribute and elements at the same time. Stuff like that doesn't make sense:

```xml
<?xml version="1.0" encoding="utf-8"?>
<container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
    <services>
        <service id="autowire_star" class="Foo" autowire="*">
           <autowire>setFoo</autowire>
            <autowire>bar</autowire>
        </service>
        <service id="autowire_array" class="Foo"  autowire="false">
            <autowire>setFoo</autowire>
            <autowire>bar</autowire>
        </service>
    </services>
</container>
```

Tobion commented:
So if you explicitly autowire a method that has no arguments it is basically the same as a normal call configuration to this method? I suspect people will then misuse autowiring just to call certain methods since there are now 2 ways to do that.

dunglas commented on src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd:
```diff
@@ -204,4 +205,10 @@
+
+  <xsd:simpleType name="boolean_or_star">
+    <xsd:restriction base="xsd:string">
+      <xsd:pattern value="(%.+%|true|false|\*)" />
```
This isn't a BC break because it was already not using the the boolean type of XML schema but the locally defined one. I've just copied, pasted and extended the current type definition:

```xml
  <xsd:simpleType name="boolean">
    <xsd:restriction base="xsd:string">
      <xsd:pattern value="(%.+%|true|false)" />
    </xsd:restriction>
  </xsd:simpleType>
```

Tobion replied:
I see, it's to allow parameters. I'm pretty sure allowing dynamic values for autowiring should not be allowed and makes no sense after all. I don't see the point of it.

dunglas replied:
Ok got it. But we cannot change it because it was already allowed since 2.8 (because it was using the common `boolean` type that allows this usage).

dunglas commented:
@Tobion thanks for the review, comments fixed now.

> So if you explicitly autowire a method that has no arguments it is basically the same as a normal call configuration to this method? I suspect people will then misuse autowiring just to call certain methods since there are now 2 ways to do that.

I think that it's not the same intent:

* When you use `<autowire>`, it means "always fill parameters of this method, if I add or remove new ones, it must not fail"
* When you use `<call>`, it's the old good call without any magic

But again, and as pointed in the doc, I would discourage end users to use setter autowiring. It's a feature designed for designers of RAD frameworks and bundles.

Tobion commented:
> When you use <autowire>, it means "always fill parameters of this method, if I add or remove some, it must not fail"

Fine for me.

One more thing. I think we should make the programmatic `Definition` methods more explicit and not use a multivalued `@param bool|string|string[] $autowired` parameter. IMO we need to distinguish between configuration and programmatic API.

```yaml
autowire: true # constructor autowiring
autowire: [__construct, setFoo, setBar] # autowire whitelisted methods only
autowire: '*' # autowire constructor + every setters (following existing rules for setters autowiring)
```

makes sense as @dunglas and @nicolas-grekas pointed out. But using a magic `*` in a programmers API like

```
$fooDefinition = new Definition('Foo');
$fooDefinition->setAutowired('*');
````
does not makse sense.

So I'd prefer to change the methods on the `Definition` class similar to what @ro0NL suggested in https://github.com/symfony/symfony/pull/19631#issuecomment-240681547.

```
$definition->setAutowired(Definition::CONSTRUCTOR|Definition::SETTERS);
$definition->setAutowiredMethods(array('setFoo', 'notASetter'));
```

dunglas commented:
It will be difficult to make this new API backward compatible with the old one. Currently it `autowire` is a boolean.

ro0NL commented:
@dunglas what about this approach?

```php
$def->setAutowired(false); // no autowiring
$def->setAutorwired(true|Def::CONSTRUCTOR); // constructor autowiring
$def->setAutowired(Def::SETTERS); // setter autowiring
$def->setAutorwired(array('foo')); // whitelist autowiring
$def->setAutorwired(Def::CONSTRUCTOR|Def::SETTERS); // combined autowiring
```

```yml
autowire: false # no autowiring
autowire: true | 'constructor' | constant('Def::CONSTURCOTR') # constructor autowiring
autowire: 'setters' | constant('Def::SETTERS') # setter autowiring
autowire: # whitelist autowiring
  - foo
autowire: # combined autowiring
  constructor: true
  setters: true
```

dunglas commented:
@ro0NL I don't want to change the format for XML and YAML because we already have reached a consensus.

Your proposal for the PHP API doesn't fix the issue raised by @Tobion: the `setAutowire` method still accept several types of parameters (`string`, `bool` and `string[]`); exactly like the current implementation.

Using a bit field isn't the right choice here: you cannot accumulate options. You can only choose one option between constructor autowiring, constructor + setter autowiring or a list of methods to autowire.

IMO the current implementation is good enough even if I would prefer accepting only one type for `setAutowire`. It's expressive, easy to learn and flexible. All loaders (PHP, XML and YAML) also have a similar API (better learning curve).

Maybe can we just introduce a `Definition::SETTER_AUTOWIRING = '*'` to avoid using a special raw string in the PHP API.

ro0NL commented:
Fair enough. Just saying im not sure `*` meaning `constructor + setters` is intuitive.. especially as we shouldnt promote setter injection.. right? This is making it more or less special.

`autowire: '*' vs. autowire: { constructor: true, setters: true }`.

I'd prefer the latter in terms of less magic, extensibility and verbose configuration.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -84,13 +87,13 @@ public static function createResourceForClass(\ReflectionClass $reflectionClass)
      * @throws RuntimeException
      */
-    private function completeDefinition($id, Definition $definition, $autowired)
+    private function completeDefinition($id, Definition $definition, array $autowired = array('__construct'))
```
since this is private, let's make the arg mandatory to make the calling code easier to read?

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -41,7 +44,7 @@ public function process(ContainerBuilder $container)
             $this->container = $container;
             foreach ($container->getDefinitions() as $id => $definition) {
-                if (false !== $autowired = $definition->getAutowired()) {
+                if (!empty($autowired = $definition->getAutowired())) {
```
I'm not a bug fan of using `empty`. I'd prefer `if ($autowired = $definition->getAutowired()) {`

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -115,50 +118,34 @@ private function completeDefinition($id, Definition $definition, $autowired)
-        if (true === $autowired && $constructor = $reflectionClass->getConstructor()) {
-            return array($reflectionClass->getConstructor());
-        }
+        $found = array();
```
not required

dunglas replied:
It is, otherwise the array may be uninitialized (if the class has no method for instance).

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -115,50 +118,34 @@ private function completeDefinition($id, Definition $definition, $autowired)
-        return $this->getSetters($reflectionClass);
-    }
+            foreach ($autowired as $methodPattern) {
+                if ($methodPattern === $reflectionMethod->name || ('*' === substr($methodPattern, -1) && 0 === strpos($reflectionMethod->name, substr($methodPattern, 0, -1)))) {
```
I suggest to turn the `$autowired` into a regexp at the beginning of the method, then using preg_match (in case insensitive mode) with e.g.:
`$autowiredRx = implode('|', array_map(function ($m) { return str_replace('\*', '.*', preg_quote($m, '/')); }, $autowired));`

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -115,50 +118,34 @@ private function completeDefinition($id, Definition $definition, $autowired)
-            if (!$reflectionMethod->isStatic() && 1 === $reflectionMethod->getNumberOfParameters() && 0 === strpos($reflectionMethod->name, 'set')) {
-                yield $reflectionMethod;
-            }
+        if (!empty($notFound = array_diff($autowired, $found))) {
```
empty not needed

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -680,15 +680,28 @@ public function getAutowired()
+            return $this;
+        }
+
+        if (true === $autowired) {
```
I think this case should be more predictable and never take the previous value into account: 
`$autowired = array('__construct')`

dunglas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -24,6 +24,9 @@
 class AutowirePass implements CompilerPassInterface
 {
+    /**
+     * @var ContainerBuilder
```
Without it PHPStorm's autocompletion doesn't work, but I can remove it :)

dunglas replied:
There is no constructor btw (it's why it's required).

nicolas-grekas replied:
oh right, the function I looked at is `process`
ok then

nicolas-grekas commented:
:+1: on my side. Ping @Tobion, your turn :)

dunglas commented:
Thanks to @nicolas-grekas's feedback, this PR has been updated to remove edge cases and simplify the implementation:

```yaml
services:
    foo:
        class: Foo
        autowire: ['__construct', 'set*'] # Autowire constructor and all setters
        autowire: true # Converted by loaders in `autowire: ['__construct']` for BC
        autowire: ['foo', 'bar'] # Autowire only `foo` and `bar` methods
```

It should resolve @Tobion's concerns, and it allows using patterns as suggested by @Ener-Getick.

fabpot commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -662,19 +662,37 @@ public function setAutowiringTypes(array $types)
      *
      * @return Definition The current instance
      */
     public function setAutowired($autowired)
```
Regarding my previous comment, `setAutowired()` made sense when it took a book as an argument, not so much now. `setAutowiring()` or `setAutowiredMethods()` sounds better to me. Just my 2cents of course, we should have more opinions here.

fabpot replied:
Having another method name also means that we can keep the current one as is, deprecate it, and remove it in 4.0 easily.

ro0NL replied:
@fabpot should it not support a boolean value? I like at least `autowire: true` for file loaders (opposed to `autowire: ['__construct']`).

fabpot replied:
I don't have any issue with `autowire: true` in YAML/XML, but do we really need that support in PHP?

fabpot commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -662,19 +662,37 @@ public function setAutowiringTypes(array $types)
-     * @param bool $autowired
+     * Allowed values:
+     *   - true: constructor autowiring
+     *   - array('__construct', 'set*', 'setBar'): autowire whitelisted methods only
```
I suppose `false` is also supported and does the same as `array()`, right?

Ener-Getick replied:
`setBar` should already be matched because `set*`. What about using `initialize`  instead of `setBar` to make this method easier to understand or use more specific patterns such as `['setApp*', 'setBar']`?

dunglas replied:
Indeed.

fabpot commented:
This one needs to be merged before 3.2 or #17608 must be reverted. I think we're almost there, I made some naming suggestions, can @symfony/deciders have a look so that we can decide either way?

nicolas-grekas commented:
> Final thoughts; my vote goes to setAutowire($boolOrArray)

Wouldn't be mine: union types require higher cyclomatic complexity. Having
simple types looks better to me.

ro0NL commented:
Perhaps should be avoided, but in this case (to me) it makes most sense from a _configuring_ perspective.

Eventually a YAML/XML/PHP definition represents the same thing, and therefore should have similar schema/API imo.

nicolas-grekas commented:
My vote would be for reverting #17608 and merge this one for 3.3.
This one has a high impact on DX, yet DX needs some time to mature IMHO, and this maybe be too fresh for 3.2.

dunglas commented:
I agree, it's too late for 3.2.

dunglas commented:
I've rebased the PR and updated method's naming.

I've added two new methods `setAutowiredMethods(array $autowiredMethods)` and `getAutowiredMethods() : array` because the naming is better and it allows to remove the union type.

However I've kept  `isAutowired() : bool` and `setAutowired(bool $autowired)` (without deprecating them) because it doesn't hurt and it looks better in term of DX when not using method autowiring and method autowiring should almost never be used when not developing a library or a framework.

ro0NL commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -97,12 +103,75 @@ private function completeDefinition($id, Definition $definition)
+        $found = array();
+        $regexList = array();
+        foreach ($autowired as $pattern) {
+            $regexList[] = '/^'.str_replace('\*', '.*', preg_quote($pattern, '/')).'$/i';
```
Perhaps use `.+`, but it depends if `set*` should match `set()`. Not sure :)

nicolas-grekas replied:
`*` comes from glob patterns, and there, `*` => `/.*/`, we shouldn't be the one redefining old semantics people are used to since years.

ro0NL commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -662,19 +662,50 @@ public function setAutowiringTypes(array $types)
+     *
+     * @return Definition The current instance
+     */
+    public function setAutowired($autowiredMethods)
```
`$autowired`

ro0NL commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -662,19 +662,50 @@ public function setAutowiringTypes(array $types)
+     */
+    public function setAutowired($autowiredMethods)
+    {
+        $this->autowiredMethods = $autowiredMethods ? array('__construct') : array();
```
I would call `setAutowiredMethods` here..

ro0NL replied:
Fair enough.

ro0NL commented on src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php:
```diff
@@ -239,6 +239,19 @@ private function parseDefinition(\DOMElement $service, $file)
+            $autowireTags[] = $type->textContent;
+        }
+
+        if (!empty($autowireTags)) {
```
`if (0 !== count($autowiredTags)) {`?

nicolas-grekas replied:
Even better: `if ($autowiredTags) {`
I'm fighting against the idea that using count() provides any type information. It does not.

ro0NL replied:
You're right.. got confused by `!` :disappointed:

Ener-Getick commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -97,12 +103,75 @@ private function completeDefinition($id, Definition $definition)
+     *
+     * @param string           $id
+     * @param \ReflectionClass $reflectionClass
+     * @param string[]         $autowired
```
`$autowiredMethods`?

Ener-Getick commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -97,12 +103,75 @@ private function completeDefinition($id, Definition $definition)
+    {
+        if ($isConstructor = $reflectionMethod->isConstructor()) {
+            $arguments = $definition->getArguments();
+        } else {
```
What about removing the `else`?
```php
$arguments = array();
if ($isConstructor = $reflectionMethod->isConstructor()) {
    $arguments = $definition->getArguments();
}

dunglas replied:
It adds an extra assignation. Not sure it worth it.

Ener-Getick replied:
As you want, that's just a détail after all ☺

Ener-Getick commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -111,7 +180,11 @@ private function completeDefinition($id, Definition $definition)
                     if (!$parameter->isOptional()) {
-                        throw new RuntimeException(sprintf('Unable to autowire argument index %d ($%s) for the service "%s". If this is an object, give it a type-hint. Otherwise, specify this argument\'s value explicitly.', $index, $parameter->name, $id));
+                        if ($isConstructor) {
+                            throw new RuntimeException(sprintf('Unable to autowire argument index %d ($%s) for the service "%s". If this is an object, give it a type-hint. Otherwise, specify this argument\'s value explicitly.', $index, $parameter->name, $id));
```
What about reversing the `if` condition here to make the exception message more readable?
```php
if (!$isConstructor) {
    return;
}

// ...

dunglas replied:
But I prefer using non-negative condition when possible (and it's possible here).

Ener-Getick commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -97,12 +103,75 @@ private function completeDefinition($id, Definition $definition)
+            $arguments = array();
+        }
+
+        $addMethodCall = false;
```
Could you add a small comment here? I don't think it's easy to understand what contains this variable without looking deeply the code.
Maybe `// Whether the method should be added to the definition as a call`.

At the same time, I don't remember if it's written somewhere in the code what's this pass behavior (e.g. only autowire methods when at least one if their arguments is not to its default value), it could be worth adding it if not.

Ener-Getick commented on src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php:
```diff
@@ -476,6 +544,37 @@ public function testIgnoreServiceWithClassNotExisting()
+        $pass = new AutowirePass();
+        $pass->process($container);
+
+        $this->assertEquals(array('Symfony\Component\DependencyInjection\Compiler\AutowirePass: Autowiring\'s patterns "not", "exist*" for service "foo" don\'t match any method.'), $container->getCompiler()->getLog());
```
`AutowirePass::class`?

Ener-Getick commented on src/Symfony/Component/DependencyInjection/Tests/DefinitionTest.php:
```diff
@@ -290,6 +290,16 @@ public function testAutowired()
         $this->assertTrue($def->isAutowired());
+        $this->assertEquals(array('__construct'), $def->getAutowiredMethods());
+
+        $def->setAutowired(array('foo'));
```
`setAutowiredMethods`?

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -97,12 +103,75 @@ private function completeDefinition($id, Definition $definition)
+        $found = array();
+        $regexList = array();
+        foreach ($autowiredMethods as $pattern) {
+            $regexList[] = '/^'.str_replace('\*', '.*', preg_quote($pattern, '/')).'$/i';
```
No need IMHO

dunglas commented:
@fabpot can I merge this one? It will help for #18193 and #20738.

nicolas-grekas commented:
:+1: for merging asap, so that we can play with the feat asap and tweak it if needed in the coming months.

fabpot commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -126,24 +199,35 @@ private function completeDefinition($id, Definition $definition)
                         } else {
-                            throw $e;
+                            // The exception code is set to 1 if the exception must be thrown even if it's a setter
+                            if (1 === $e->getCode() || $isConstructor) {
```
Can we use a specific getter/setter instead of a magic 1 value for the code?

dunglas replied:
Done (but I find the use of a specific code less intrusive than using accessors, here we add methods to a generic exception for an implementation detail of the `AutowirePass`).

fabpot commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -662,19 +662,50 @@ public function setAutowiringTypes(array $types)
      *
      * @return Definition The current instance
      */
     public function setAutowired($autowired)
```
This one could be deprecated, right?

dunglas replied:
I propose to deprecate it in another PR (if it's worth it) because we need determine how we deal with symfony/symfony#20648

fabpot commented on src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php:
```diff
@@ -648,9 +747,74 @@ public function setBar(Bar $bar)
+{
+    public function setFoo(Foo $foo)
+    {
+        // should be called
```
What about throwing an exception here instead of the comment?

dunglas replied:
This setter should be called (and shouldn't throw anything).

fabpot replied:
nevermind :)

fabpot commented:
Thank you @dunglas.
